### ISSUE TITLE
Improve onboarding experience

### ISSUE THREAD

**Issue by larrysalibra** [blockstack/blockstack-browser#534]

**Comment by shea256** [comment]
# Landing

![blockstack - landing](https://user-images.githubusercontent.com/994646/28085434-bce1c6fc-6649-11e7-83ce-ec67e571a412.png)

# Restore - Page 1

![blockstack - restore 1](https://user-images.githubusercontent.com/994646/28085455-c7fecef4-6649-11e7-898a-37224720d56e.png)

# Register - Page 1

![blockstack - signup 1](https://user-images.githubusercontent.com/994646/28085439-c1a09c86-6649-11e7-9245-07050f439a3b.png)

# Register - Page 2

![blockstack - signup 2](https://user-images.githubusercontent.com/994646/28085441-c1ac25c4-6649-11e7-9e82-090e295efb72.png)

# Register - Page 3

![blockstack - signup 3](https://user-images.githubusercontent.com/994646/28085443-c1ae101e-6649-11e7-8cc4-77a2ef2d20c3.png)

# Register - Page 4

![blockstack - signup 4](https://user-images.githubusercontent.com/994646/28085440-c1abccdc-6649-11e7-840c-55cbd855bc51.png)

# Register - Page 5

![blockstack - signup 5](https://user-images.githubusercontent.com/994646/28085442-c1acdd84-6649-11e7-98cf-f14819ae8fcd.png)

# Register - Page 6

![blockstack - signup 6](https://user-images.githubusercontent.com/994646/28085444-c1c11826-6649-11e7-95c3-f2daa4256bb9.png)

**Comment by guylepage3** [comment]
closing issue.

**Comment by larrysalibra** [comment]
@shea256 @guylepage3 For page 5: It seems like it would be a better experience to do something like copay and provide the users with boxes that include the 12 words in scrambled order and have them click/tap on each word to type their identity key instead of having to type each letter? 

For email notifications, is this intended as a way to implement notifying users when their name is ready to use? 
https://github.com/blockstack/blockstack-browser/issues/565

**Comment by guylepage3** [comment]
>For page 5: It seems like it would be a better experience to do something like copay and provide the users with boxes that include the 12 words in scrambled order and have them click/tap on each word to type their identity key instead of having to type each letter?

@larrysalibra this adds quite a bit of friction on the user's part. I feel that's the best route to go. Having said that, we should have a visual like a pad of paper and pen with an alert icon of some kind so that the user understands that this is important that they write this down.

**Comment by larrysalibra** [comment]
@guylepage3 
> this adds quite a bit of friction on the user's part. I feel that's the best route to go.

Can't tell if you're agreeing or disagreeing with me.

**Comment by guylepage3** [comment]
>Can't tell if you're agreeing or disagreeing with me.

@larrysalibra your method adds more friction so therefore not agreeing with your method.

**Comment by larrysalibra** [comment]
@guylepage3 have you tried word based identity entry on copay? I found it a lot less work and lower friction than typing out a bunch of words. I actually thought "wow - this is feels like magic" 

* It removes the potential to make typos. 
* 12 clicks versus ~60 letter presses
* No "i'm sorry your identity key doesn't match" and having to scan 60 letters to figure out which letter you missed or transposed - it's obvious if you put the words in the wrong order.

Anyways, I'm fine with the typing way - it's easier to implement...just found the other way a much better experience which is why I brought it up. :-)

**Comment by guylepage3** [comment]
>@guylepage3 have you tried word based identity entry on copay? I found it a lot less work and lower friction than typing out a bunch of words. I actually thought "wow - this is feels like magic"

@larrysalibra I believe what are trying to get the user to do is to not type anything out but to get them to store their pass phrase on physical paper for security purposes. So therefore adding more steps to get their pass phrase is more friction and more work.

**Comment by larrysalibra** [comment]
@guylepage3 page 5 is asking the user to read the identity key they've written on paper and type it back in again to make sure they wrote it down properly. The method I'm suggesting doesn't add any steps. It **replaces** the typing ~60 characters in page 5 with clicking 12 words instead.  

Here are some screenshots:

![img_4162](https://user-images.githubusercontent.com/597182/28137695-b58a75e2-6780-11e7-9cbc-53a0d20ce7bf.PNG)

![img_4163](https://user-images.githubusercontent.com/597182/28137704-bd0ead7e-6780-11e7-9d81-d7ffa1cbd885.PNG)
 
![img_4164](https://user-images.githubusercontent.com/597182/28137712-c4ea0bce-6780-11e7-9022-118dfe376e13.PNG)

(note: I didn't write down this test phrase down so I was just randomly entering words in the last screenshot)

**Comment by guylepage3** [comment]
@larrysalibra my bad. i could not tell which screen was which. i was looking at **screen 4**. yes. what you're proposing is a much better experience for verifying a phrase being written down. 

However, I question the actual process of verifying a written down identity key in the first place with out emphasizing the importance of the password as so many users will forget their password let alone their passphrase. Unless we're [#606: removing the password](https://github.com/blockstack/blockstack-browser/issues/606). Which would be cool.

**Register Page 4**
1. Change headline to read, "Write down your identity key for account recovery — Keep it in a safe place & don't show it to anyone"

**Register Page 5**
1. We will need to ensure that users can go back when they decide not to write down their identity key.

**Register Page 6**
1. Change headline to read, "Enter your email address to get important information about your identity and/or account"
2. Change "Finish" button to "Create account"

**Comment by shea256** [comment]
@larrysalibra:

> For page 5: It seems like it would be a better experience to do something like copay and provide the users with boxes that include the 12 words in scrambled order and have them click/tap on each word to type their identity key instead of having to type each letter?

Yes I do believe this would be a much better experience. I was just worried about the complexity of implementation. Are we game for this? If so let's do it. I've seen another way of doing it which essentially asks the user the following: "enter the 2nd word and the 7th word".

> For email notifications, is this intended as a way to implement notifying users when their name is ready to use?

Yes. This will be for notifying the user when their name is ready to use, when they received bitcoin, when bitcoin leaves their account, when someone follows them, etc.

@guylepage3:

> Change headline to read, "Write down your identity key for account recovery — Keep it in a safe place & don't show it to anyone"

Yes I wanted to say this but I wanted to keep every phrase to 3 lines max.

> We will need to ensure that users can go back when they decide not to write down their identity key.

Yes great call. Every page needs a back button.

> Change headline to read, "Enter your email address to get important information about your identity and/or account"

I wouldn't phrase it like this because it sounds like you're getting added to a mailing list.

> Change "Finish" button to "Create account"

Technically the account was already partially created. I'm on the fence on this one. Will think more about the right word.

**Comment by larrysalibra** [comment]
From our call today, we decided to add the password requirement back in (makes #606 invalid).

**Comment by shea256** [comment]
Exactly what we have here: https://github.com/blockstack/blockstack-browser/issues/534#issuecomment-314541398

**Comment by shea256** [comment]
We won't have a password requirement in the onboarding process, but then we'll nudge the user to set a password in the settings page.

**Comment by larrysalibra** [comment]
We're going to push off the copay-style click to verify your identity key until later.

**Comment by guylepage3** [comment]
@shea256 @larrysalibra can we close out this issue and lock down deliverables for on-boarding?

**Comment by shea256** [comment]
@guylepage3 You mean create a new issue and start from scratch? Most of the relevant info is here: https://github.com/blockstack/blockstack-browser/issues/534#issuecomment-314541398.

**Comment by guylepage3** [comment]
I mean. this is complete. I have other issues open for the build and the design.

**Comment by shea256** [comment]
What has been completed here? Can I use the on-boarding experience now? Can you link to the other issues please? Also let's wait on @larrysalibra before deciding to close. It seems like this issue won't be done until someone can try the on-boarding end-to-end.

**Comment by shea256** [comment]
Just went through the on-boarding experience. Great work on this @larrysalibra @guylepage3!

Jotting down some notes on the process here for improvements:

# 1. Figure out how to display the modal in a way that works well with small screens.

There's currently overlap on the navbar and that looks a bit strange. Of course when we switch over to a vertically oriented pane for the whole app, this problem will go away.

<img width="1232" alt="screen shot 2017-07-23 at 10 32 47 am" src="https://user-images.githubusercontent.com/994646/28500265-03bd6322-6f93-11e7-9e09-8f7f70d73c43.png">

# 2. Use icons that more effectively embody the steps in the onboarding process.

I have a couple of icons above that work pretty well. Happy to send those over to you.

<img width="432" alt="screen shot 2017-07-23 at 10 33 27 am" src="https://user-images.githubusercontent.com/994646/28500279-67eed506-6f93-11e7-9dea-a47ed4f95dc1.png">
<img width="433" alt="screen shot 2017-07-23 at 10 33 19 am" src="https://user-images.githubusercontent.com/994646/28500280-67f1a272-6f93-11e7-8a92-cfbf34d1aea6.png">

# 3. Switch the pages for password entry and the identity key explanation

- Page 1
  - Blockstack has no 3rd parties: an identity key on your device gives you access
  - Image
  - Button: Continue
- Page 2
  - Choose a password to protect your identity key
  - Create a secured identity key by choosing a password
  - Field label: Password, Field placeholder: Password
  - Field label: Confirm Password, Field placeholder: Confirm Password
  - Button: Create identity key

<img width="439" alt="screen shot 2017-07-23 at 10 33 43 am" src="https://user-images.githubusercontent.com/994646/28500327-693cf7c0-6f94-11e7-94f4-8ab0c91f46e7.png">
<img width="433" alt="screen shot 2017-07-23 at 10 34 30 am" src="https://user-images.githubusercontent.com/994646/28500328-693e8518-6f94-11e7-9954-ca325e942faa.png">

# 4. Update the copy for the email page

- Enter your email address to get useful notification regarding your account
- Field label: Email Address, Field placeholder: Email Address
- Finish

# 5. Only allow the mail page to close when one of the buttons is clicked

Accidentally clicking outside the email page will cause it to close. Also if you click to the left or right of the "Skip" button the email page will close. Neither of these things should happen.

There were two times I went through the onboarding process where I accidentally clicked somewhere and I never got to finish entering my email. Also, if I don't enter it there, I can never enter it anywhere again, so it's even more consequential.

# 6. A storage provider is required for auth so add it to the on-boarding process

It's a much more jarring experience to the user to try and sign in somewhere only to realize that they haven't been fully set up in the on-boarding process and they need to do something else. We need to give users the option to connect a storage provider in the on-boarding until we allow users to use a default storage provider. We can still give them a skip button. Happy to provide more details on this page.

**Comment by guylepage3** [comment]
@larrysalibra @shea256 

I can resolve issue 1. It just needs styles to be defined for all 4 media queries.

**Comment by larrysalibra** [comment]
Quick recap on @shea256 's comments:

**Point 1** 
@guylepage3  will handle this. Here's the tracking issue: #650 

**Point 2**
@guylepage3 @shea256 please coordinate on this. tracking issue: #651

**Point 3**
Done and merged into the `develop` branch. Please try it out!

**Point 4**
Done and merged into the `develop` branch. Please try it out!

**Point 5**
Email page isn't entirely implemented. (Awaiting server side changes #648) I updated it so that it will require you to enter an email before clicking Finish makes it disappear. This is merged into `develop`.

As to the problem with clicking on whitespace making the modal disappear, that's a style issue that affects multiple on-boarding views: @guylepage3 can you take a look? tracking issue here: #643 

**Point 6** 

I created an issue for adding storage back into the on-boarding flow: #645 please add any thoughts you have about that to that issue. 


Thanks again for the feedback!

I'm going to close out this issue. Please create individual issues for additional improvements to the on-boarding process.

**Comment by shea256** [comment]
Awesome, fast turnaround on these!

Just made this comment here regarding (3): https://github.com/blockstack/blockstack-browser/commit/8599dc7a6bc203c425997851bb296954c1f43d96#commitcomment-23251612. Happy to open another issue regarding this if you need me to.
